### PR TITLE
Streaming final agent response callbacks

### PR DESCRIPTION
**Pull request by thaiminhpv**
This is a comprehensive version of callbacks when we want to streaming the only the last response of the agent.

This fix issue #2483 but in a more comprehensive way. This is NOT duplicate of #4630 .

I know there is a great work has been done by @UmerHA for **FinalStreamingStdOutCallbackHandler** class, and this PR work is inspired by him, but I want that class to be a simple use case for final streaming. What this PR does is a comprehensive version of final streaming with the following additional feature:

- **Multiple answer prefix phrases**: Just like OpenAI they allows multiple stop sequences, this class can detect multiple stop sequences before it begin streaming.
- **Post-process on-the-fly**: This utilize the above algorithm (matching multiple prefix phrases) so I think this is suitable to place it here.
- **Easy-to-use Callback** on each new token **during streaming of final agent response**.
- **Abnormal streaming detection**: if the streaming contains unexpected tokens generated by model, then raise an Exception.

#### Quick setup

```python
from langchain.agents import load_tools, initialize_agent, AgentType
from langchain.llms import OpenAI
from langchain.callbacks import StreamingLastResponseCallbackHandler

llm = OpenAI(temperature=0, streaming=True)
tools = load_tools(["serpapi", "llm-math"], llm=llm)
agent = initialize_agent(tools, llm, agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose=False)
```

#### Usage 1 - use as callback function for **last response new token**
```python
stream = StreamingLastResponseCallbackHandler.from_agent_type(agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION)
@stream.on_last_response_new_token()
def on_new_token(token: str):
    if token is StopIteration:
        print("\n[Done]")
        return
    else:
        print(token, end="", flush=True)

agent.run("Who is Leo DiCaprio's girlfriend? What is her current age raised to the 0.43 power?", callbacks=[stream])
```
#### Usage 2 - use as iterator
```python
stream = StreamingLastResponseCallbackHandler.from_agent_type(agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION)
def _run():
    agent.run("Who is Leo DiCaprio's girlfriend? What is her current age raised to the 0.43 power?", callbacks=[stream])
threading.Thread(target=_run).start()

for token in stream:
    print(token, end="", flush=True)
```

#### Usage 3 - Post process on-the-fly
```python
import tiktoken
enc = tiktoken.get_encoding("cl100k_base")
stream = StreamingLastResponseCallbackHandler.from_agent_type(agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION)

@stream.postprocess(sliding_window_step=1, window_size=3)
def postprocess_func(tokens: List[str]) -> List[str]:
    sentence = "".join(tokens).replace("LLM", "LangChain")
    words = [enc.decode([t]) for t in enc.encode(sentence)]  # postprocess output can have different size!
    return words

def _run():
    agent.run("Say 'Large Language Model (LLM) is great!'", callbacks=[stream])
threading.Thread(target=_run).start()

for token in stream:
    print(token, end="", flush=True)
```

#### Who can review?

Hi @hwchase17 and @agola11, please review my PR, I do have an usage example in the docstring of this **StreamingLastResponseCallbackHandler** class.

I appreciate any feedback, please review my PR soon.

P/s: Updated using decorator approach.

### CODE REVIEW

**Comment by UmerHA** [comment]
@thaiminhpv cool! Especially love the postprocessing on the fly.
One question: In `stream = StreamingLastResponseCallbackHandler.from_agent_type(agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION)`, why do we need the agent type to initialize this callback?

**Comment by UmerHA** [comment]
@thaiminhpv Also a suggestion: Consider renaming the class to make this PR backward-compatible. If people are already using `FinalStreamingStdOutCallbackHandler`, they'll get errors.

**Comment by thaiminhpv** [comment]
@UmerHA thank you!

> One question: In `stream = StreamingLastResponseCallbackHandler.from_agent_type(agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION)`, why do we need the agent type to initialize this callback?

It is just a convenient shorthand for `StreamingLastResponseCallbackHandler(answer_prefix_phrases=["Final Answer:"])`

We can use this class by custom defining `answer_prefix_phrases` like this

```python
stream = StreamingLastResponseCallbackHandler(answer_prefix_phrases=["Now I know the final answer:"])
```

As in here:

https://github.com/thaiminhpv/langchain/blob/e208414a9ae3cc9a6d84b6ebfe5b2c7e4c0cea77/langchain/callbacks/streaming_last_response_callback.py#L308-L323

---

> Also a suggestion: Consider renaming the class to make this PR backward-compatible. If people are already using FinalStreamingStdOutCallbackHandler, they'll get errors.

@UmerHA  IMO, I prefer to keep both classes as separate files, as each classes serve different purposes.

File: langchain/callbacks/streaming_last_response_callback.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+
+        try:
+            import tiktoken
+        except ImportError:
+            raise ImportError(
+                "Could not import tiktoken python package. "
+                "This is needed in order to calculate detection_windows_size for StreamingLastResponseCallbackHandler"
+                "Please install it with `pip install tiktoken`."
+            )
+        self._enc = tiktoken.get_encoding(tiktoken_encoding)
```
Comment (baskaryan) [inline langchain/callbacks/streaming_last_response_callback.py:115]: is it worth letting folks pass in whatever tokenize they want (can still default to tiktoken "cl100k_base")?

File: docs/extras/modules/agents/how_to/streaming_last_agent_response.ipynb
Comment (baskaryan) [inline docs/extras/modules/agents/how_to/streaming_last_agent_response.ipynb:1]: thanks for pr @thaiminhpv! could we add a not explaining the difference between this and FinalStreamingStdOutCallbackHandler? imagine it might otherwise confuse a lot of folks

File: langchain/callbacks/streaming_last_response_callback.py (hunk 1 above)
Comment (thaiminhpv) [reply langchain/callbacks/streaming_last_response_callback.py]: > is it worth letting folks pass in whatever tokenize they want (can still default to tiktoken "cl100k_base")?

@baskaryan Thank you for the suggestion, I have implemented it, please check it!

https://github.com/thaiminhpv/langchain/blob/4c8bdceecf6a8eef06968dadea06878bc4875676/libs/langchain/langchain/callbacks/streaming_last_response_callback.py#L106-L135

File: docs/extras/modules/agents/how_to/streaming_last_agent_response.ipynb
Comment (thaiminhpv) [reply docs/extras/modules/agents/how_to/streaming_last_agent_response.ipynb:1]: @baskaryan 
I think the differences are as follows:

The `FinalStreamingStdOutCallbackHandler`, as the name suggests, it just output the final response of the agent to stdout. User can implement anything such as postprocess or complex logic on top of that.

The `StreamingLastResponseCallbackHandler`, is like the **battery included** version of `FinalStreamingStdOutCallbackHandler`, with additional features that I mentioned earlier in this PR.

In other words:

- If the user needs to custom some complex logic themselves -> go for `FinalStreamingStdOutCallbackHandler`

- If user want to post-process on-the-fly, or use multiple answer-prefix-phrase, or use abnormal detection feature -> go for `FinalStreamingStdOutCallbackHandler`.

I'm not sure where to document the explanation of the difference. Do you have any idea?

**Comment by thaiminhpv** [comment]
BTW, I've rebased my branch (manually) to be ahead of the `master` branch, following by a force push.

This means we can merge my branch without any conflict, for now.

**Comment by leo-gan** [comment]
@baskaryan Could you, please, review it?

**Comment by hinetabi** [comment]
This PR is so helpful, please merge it. I want this in new version of langchain.

**Comment by ilianherzi** [comment]
+2 would love to see this!

**Comment by IdkwhatImD0ing** [comment]
Is there a way to make this work for openai functions?

**Comment by leo-gan** [comment]
@baskaryan FYI

**Comment by thaiminhpv** [comment]
This pull request was created prior to the introduction of OpenAI functions and the LangChain Expression Language (LCEL). I think this pull request is now obsolete and outdated.

**Comment by alexgg278** [comment]
Isn't it possible to adapt it to LCEL and OpenAI functions? It would be usefult to have it

**Comment by Tesax123** [comment]
+1, this would be very useful.
